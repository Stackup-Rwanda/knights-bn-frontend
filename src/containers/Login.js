import React, { Component } from 'react';
import '../assets/styles/containers/landingPage.scss';


class Login extends Component {
  render() {
    return (
      <div className="background">
        <h2>Barefoot Nomad Login Page</h2>

      </div>
    );
  }
}
export default Login;
