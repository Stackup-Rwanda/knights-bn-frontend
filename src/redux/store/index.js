import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
// import logger from 'redux-logger';
import rootReduce from '../reducers/index';

const store = createStore(
  rootReduce,
  applyMiddleware(thunk),
);

export default store;
