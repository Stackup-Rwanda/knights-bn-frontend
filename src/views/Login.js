import React, { Component } from 'react';
import { Form, Spinner } from 'react-bootstrap';
import '../assets/styles/containers/loginPage.scss';
import { connect } from 'react-redux';
import {
  fetchUsers,
  validateInput,
  loginUsersSuccess,
  handleError,
  setLoadingStatus,
} from '../redux/actions/index';
import { Redirect } from 'react-router-dom';
import TextBox from '../components/Textbox';
import Button from '../components/Button';
import Line from '../components/line';
import { thunk } from '../redux/thunk/index';
import { withRouter } from 'react-router';
import AlertComponent from '../components/Alert';
import SocialLogin from '../components/SocialLogin';

export class Login extends Component {
  state = {
    data: { email: '', password: '' },
    error: ' ',
    loggedIn: false,
  };

  login = async (e) => {
    e.preventDefault();
    this.props.handleError('');
    const { validations: validatedFields } = this.props;
    const invalidFields = Object.keys(validatedFields).filter(
      (field) => validatedFields[field] === 'is-invalid'
    );
    let validatedFieldsToArray = Object.keys(validatedFields);
    if (validatedFieldsToArray.length > 1 && invalidFields.length === 0) {
      this.props.setLoadingStatus(true);
      await this.props.thunk(
        'post',
        '/auth/login',
        loginUsersSuccess,
        this.props.user
      );
      this.props.setLoadingStatus(false);

      const { isLoggedIn } = this.props.userData;
      isLoggedIn
        ? this.props.history.push('/home')
        : this.props.handleError(
            'Incorrect email or password. Please try again.'
          );
    } else {
      const errorMsg =
        validatedFieldsToArray.length === 0
          ? 'Please fill in the required fields.'
          : this.props.validations.email === 'is-valid' &&
            !this.props.validations.password
          ? 'Please enter your password.'
          : 'Wrong email address.';
      this.props.handleError(errorMsg);
    }
  };

  render() {
    const { validateInput, error, isLoading } = this.props;
    return (
      <div className="login-page loginContainer">
        <Spinner
          animation="border"
          variant="primary"
          className={isLoading ? 'spinner--position__center' : 'hide'}
        />
        <Form className="form">
          <h1>Sign in</h1>
          <AlertComponent isError={error ? true : false} errorMsg={error} />
          <p>Login with social media</p>
          <SocialLogin
            googleAction="Sign in with Google"
            facebookAction="Sign in with Facebook"
          />
          <div id="or">
            <Line className="Line" />
            <div id="or_">
              <span>or</span>
            </div>
            <Line className="Line_" />
          </div>
          <TextBox
            type="email"
            placeholder="email"
            id="email"
            name="email"
            onChange={validateInput}
            label="email"
            value={this.props.user.email}
          />
          <TextBox
            type="password"
            placeholder="password"
            id="password"
            name="password"
            onChange={validateInput}
            label="password"
            value={this.props.user.password}
          />
          <p>{this.props.token ? <Redirect to="home" /> : null}</p>
          <Button
            aria-label="login"
            onClick={(e) => this.login(e)}
            id="loginBtn"
            label="Login"
          />
          <a id="forgotPassword" href="#">
            <p>Forgot Password?</p>
          </a>
        </Form>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    token: state.user.token,
    error: state.errorHandler.error,
    validations: state.eventHandler.validations,
    user: state.eventHandler.user,
    userData: state.user.data,
    isLoading: state.eventHandler.isLoading,
  };
};

const mapDispatchToProps = {
  fetchUsers,
  validateInput,
  thunk,
  handleError,
  setLoadingStatus,
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Login));
