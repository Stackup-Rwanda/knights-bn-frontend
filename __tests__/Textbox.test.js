import { render, cleanup, waitFor } from '@testing-library/react';
import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import userEvent from '@testing-library/user-event';
import Login from '../src/views/Login';
import store from '../src/redux/store';

const LoginComponent = () => render(
  <Provider store={store}>
    <Router>
      <Login />
    </Router>
  </Provider>,
);
describe('=============>LOGIN<=============', () => {
  afterEach(cleanup);
  it('should render login component', () => {
    const { asFragment } = LoginComponent();
    expect(asFragment(<Login />)).toMatchSnapshot();
  });
  it('should print error if email is not provided', () => {
    const { getByLabelText, getByText } = LoginComponent();
    const button = getByLabelText('Login');
    const password = getByLabelText('password');
    userEvent.type(password, '12345six');

    expect(password.value).toBe('12345six');
    userEvent.click(button);
    waitFor(() => expect(getByText('Wrong email address.')).toBeTruthy());
  });

  it('should print error if password is not provided', () => {
    const { getByLabelText, getByText } = LoginComponent();
    const email = getByLabelText('email');
    const button = getByLabelText('Login');

    userEvent.type(email, 'knights@gmail.com');
    expect(email.value).toBe('knights@gmail.com');
    userEvent.click(button);
    waitFor(() => expect(getByText('Please enter your password.')).toBeTruthy());
  });

  it('should print error if password is not provided', () => {
    const { getByLabelText, getByText } = LoginComponent();
    const email = getByLabelText('email');
    const button = getByLabelText('Login');
    const password = getByLabelText('password');
    userEvent.type(password, '12345six');

    userEvent.type(email, 'knights@gmail.com');
    userEvent.click(button);
    waitFor(() => expect(
      getByText('Incorrect email or password. Please try again.'),
    ).toBeTruthy());
  });

  it('should print error if password is not provided', () => {
    const { getByLabelText, getByText } = LoginComponent();
    const button = getByLabelText('Login');
    userEvent.click(button);
    waitFor(() => expect(getByText('Please fill in the required fields.')).toBeTruthy());
  });

  it('should print error if password is not provided', () => {
    const { getByLabelText, getByText } = LoginComponent();
    const email = getByLabelText('email');
    const button = getByLabelText('Login');
    const password = getByLabelText('password');

    userEvent.type(password, 'Niyonkuru@1');
    userEvent.type(email, 'alain.maxime@gmail.com');

    userEvent.click(button);
    waitFor(() => expect(getByText('Home Page')).toBeTruthy());
  });
});
