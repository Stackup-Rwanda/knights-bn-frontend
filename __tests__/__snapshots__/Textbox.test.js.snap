// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`=============>LOGIN<============= should render login component 1`] = `
<DocumentFragment>
  <div
    class="login-page loginContainer"
  >
    <div
      class="hide spinner-border text-primary"
    />
    <form
      class="form"
    >
      <h1>
        Sign in
      </h1>
      <span>
        <p
          class="hide"
          id="loginError"
        >
          <img
            class="alertIcon"
            src="[object Object]"
          />
        </p>
      </span>
      <p>
        Login with social media
      </p>
      <div>
        <div>
          <svg
            class="mr-2 color"
            enable-background="new 0 0 48 48"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            version="1.1"
            viewBox="0 0 48 48"
            width="1em"
            x="0px"
            xmlns="http://www.w3.org/2000/svg"
            y="0px"
          >
            <path
              d="M43.611,20.083H42V20H24v8h11.303c-1.649,4.657-6.08,8-11.303,8c-6.627,0-12-5.373-12-12
	c0-6.627,5.373-12,12-12c3.059,0,5.842,1.154,7.961,3.039l5.657-5.657C34.046,6.053,29.268,4,24,4C12.955,4,4,12.955,4,24
	c0,11.045,8.955,20,20,20c11.045,0,20-8.955,20-20C44,22.659,43.862,21.35,43.611,20.083z"
              fill="#FFC107"
            />
            <path
              d="M6.306,14.691l6.571,4.819C14.655,15.108,18.961,12,24,12c3.059,0,5.842,1.154,7.961,3.039l5.657-5.657
	C34.046,6.053,29.268,4,24,4C16.318,4,9.656,8.337,6.306,14.691z"
              fill="#FF3D00"
            />
            <path
              d="M24,44c5.166,0,9.86-1.977,13.409-5.192l-6.19-5.238C29.211,35.091,26.715,36,24,36
	c-5.202,0-9.619-3.317-11.283-7.946l-6.522,5.025C9.505,39.556,16.227,44,24,44z"
              fill="#4CAF50"
            />
            <path
              d="M43.611,20.083H42V20H24v8h11.303c-0.792,2.237-2.231,4.166-4.087,5.571
	c0.001-0.001,0.002-0.001,0.003-0.002l6.19,5.238C36.971,39.205,44,34,44,24C44,22.659,43.862,21.35,43.611,20.083z"
              fill="#1976D2"
            />
          </svg>
          <button
            class="btnStyle"
          >
            Sign in with Google
          </button>
        </div>
        <div>
          <form
            action="https://knights-bn-backnd.herokuapp.com/api/v1/auth/login/facebook"
            class="social-login-btn"
          >
            <svg
              class="mr-2 ml-3 text-light bg-info"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 320 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M279.14 288l14.22-92.66h-88.91v-60.13c0-25.35 12.42-50.06 52.24-50.06h40.42V6.26S260.43 0 225.36 0c-73.22 0-121.08 44.38-121.08 124.72v70.62H22.89V288h81.39v224h100.17V288z"
              />
            </svg>
            <button
              class="btnStyle"
            >
              Sign in with Facebook
            </button>
          </form>
        </div>
      </div>
    </form>
    <div
      id="or"
    >
      <svg
        class="Line"
        viewBox="0 1 120.5 1"
      >
        <path
          d="M 0 1 L 120.5 1"
          fill="transparent"
          id="Line"
          linecap="square"
          linejoin="miter"
          miterlimit="10"
          rendering="auto"
          stroke="rgba(231,231,231,1)"
          width="1px"
        />
      </svg>
      <div
        id="or_"
      >
        <span>
          or
        </span>
      </div>
      <svg
        class="Line_"
        viewBox="0 1 120.5 1"
      >
        <path
          d="M 0 1 L 120.5 1"
          fill="transparent"
          id="Line"
          linecap="square"
          linejoin="miter"
          miterlimit="10"
          rendering="auto"
          stroke="rgba(231,231,231,1)"
          width="1px"
        />
      </svg>
    </div>
    <div
      class="form-group"
    >
      <input
        aria-label="email"
        class="form-control"
        id="email"
        name="email"
        placeholder="email"
        type="email"
        value=""
      />
    </div>
    <div
      class="form-group"
    >
      <input
        aria-label="password"
        class="form-control"
        id="password"
        name="password"
        placeholder="password"
        type="password"
        value=""
      />
    </div>
    <p />
    <button
      aria-label="Login"
      class="btn btn-primary"
      id="loginBtn"
      label="Login"
      type="button"
    >
      Login
    </button>
    <a
      href="#"
      id="forgotPassword"
    >
      <p>
        Forgot Password?
      </p>
    </a>
  </div>
</DocumentFragment>
`;
